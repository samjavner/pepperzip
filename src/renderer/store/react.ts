import * as React from "react";
import { Commands, Dispatch, Store, Update, UseStoreOptions } from "./";

export interface Msg<M extends object, U extends Update<M>> {
    type: keyof U;
    payload?: any;
}

export function useLocalStore(log: (msg: any) => void = () => ({})) {
    return function actual<
        M extends object,
        U extends Update<M>,
        C extends Commands = {}
    >({
        init,
        update,
        createCommands,
        memo = [],
    }: UseStoreOptions<M, U, C>): Store<M, U, C> {
        const [model, reducerDispatch] = React.useReducer(
            (state: M, action: Msg<M, U>) =>
                update[action.type](state, action.payload),
            init
        );

        const vm = React.useMemo(
            () => {
                const dispatch: Dispatch<U> = {} as any;
                Object.keys(update).forEach(type => {
                    (dispatch as any)[type] = (payload: any) => {
                        const msg =
                            update[type].length === 1
                                ? { type }
                                : { type, payload };
                        log(msg);
                        reducerDispatch(msg);
                    };
                });
                const commands: C = createCommands
                    ? createCommands(model, dispatch)
                    : (({} as any) as C);
                return {
                    ...model,
                    ...dispatch,
                    ...commands,
                };
            },
            [model, update, ...memo]
        );

        return vm;
    };
}
