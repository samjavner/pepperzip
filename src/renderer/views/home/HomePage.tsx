import * as React from "react";

export const HomePage: React.FunctionComponent = () => <div>Home</div>;

export default HomePage;
