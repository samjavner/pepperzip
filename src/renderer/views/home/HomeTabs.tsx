import { Button, ButtonGroup } from "@blueprintjs/core";
import classNames from "classnames";
import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import * as actions from "../../app/actions";
import { State } from "../../app/state";

function mapStateToProps(state: State) {
    return {
        activePage: state.activeHomePage,
    };
}

function mapDispatchToProps(dispatch: Dispatch<actions.Action>) {
    return {
        onHomeClick: () => dispatch(actions.setActiveHomePage("home")),
    };
}

type HomeTabsProps = ReturnType<typeof mapStateToProps> &
    ReturnType<typeof mapDispatchToProps>;

export const HomeTabs: React.SFC<HomeTabsProps> = props => (
    <>
        <div className="pz-navtabs">
            <ButtonGroup minimal={true}>
                <Button
                    className={classNames({
                        "pz-navtabs-button": true,
                        "pz-navtabs-button--active":
                            props.activePage === "home",
                    })}
                    icon="home"
                    onClick={props.onHomeClick}
                >
                    Home
                </Button>
            </ButtonGroup>
        </div>
        <div style={{ borderTop: "1px solid lightgray" }} />
    </>
);

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(HomeTabs);
