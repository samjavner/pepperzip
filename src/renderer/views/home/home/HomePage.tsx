import * as React from "react";
import { connect } from "react-redux";
import { GlobalState } from "../../../core/model";

export const HomePage: React.SFC = props => <div>Home</div>;

const mapStateToProps = (state: GlobalState) => ({});

export default connect(mapStateToProps)(HomePage);
