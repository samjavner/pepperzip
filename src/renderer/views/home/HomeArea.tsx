import * as React from "react";
import { connect } from "react-redux";
import { GlobalState } from "../../core/model";
import * as navigation from "../../core/navigation";
import HomePage from "./home/HomePage";
import HomeTabs from "./HomeTabs";

interface HomeAreaProps {
    activePage: navigation.HomePage;
}

const HomeArea: React.SFC<HomeAreaProps> = props => (
    <>
        <HomeTabs />
        {props.activePage === "home" && <HomePage />}
    </>
);

const mapStateToProps = (state: GlobalState) => ({
    activePage: navigation.selectors.getActiveHomePage(state),
});

export default connect(mapStateToProps)(HomeArea);
