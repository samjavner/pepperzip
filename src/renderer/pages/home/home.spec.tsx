import * as React from "react";
import * as renderer from "react-test-renderer";
import * as home from "./home";

describe("home", () => {
    describe("Page", () => {
        it("renders the text 'Home'", () => {
            const component = renderer.create(<home.Page />);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});
