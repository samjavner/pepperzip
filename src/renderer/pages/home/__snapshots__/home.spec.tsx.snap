// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home Page renders the text 'Home' 1`] = `
<div>
  Home
</div>
`;
